% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adjustBins.R
\name{adjustBins}
\alias{adjustBins}
\title{Make variable-width bins}
\usage{
adjustBins(bins, reads)
}
\arguments{
\item{reads}{A \code{\link{GRanges-class}} with reads. See
\code{\link{bam2GRanges}} and \code{\link{bed2GRanges}}.}

\item{binsizes}{A vector with binsizes. Resulting bins will be close to the
specified binsizes.}

\item{stepsizes}{A vector of step sizes in base pairs, the same length as
\code{binsizes}.}

\item{chromosomes}{A subset of chromosomes for which the bins are generated.}
}
\value{
A \code{list()} of \code{\link{GRanges-class}} objects with
  variable-width bins. If \code{stepsizes} is specified, a \code{list()} of
  \code{\link{GRangesList}} objects with one entry per step.
}
\description{
Make variable-width bins based on a reference BAM file. This can be a
simulated file (produced by \code{\link{simulateReads}} and aligned with
your favourite aligner) or a real reference.
}
\details{
Variable-width bins are produced by first binning the reference BAM file
with fixed-width bins and selecting the desired number of reads per bin as
the (non-zero) maximum of the histogram. A new set of bins is then generated
such that every bin contains the desired number of reads.
}
\examples{
## Get an example BED file with single-cell-sequencing reads
bedfile <- system.file("extdata", "KK150311_VI_07.bam.bed.gz", package="AneuFinderData")
## Read the file into a GRanges object
reads <- bed2GRanges(bedfile, assembly='mm10', chromosomes=c(1:19,'X','Y'),
                     min.mapq=10, remove.duplicate.reads=TRUE)
## Make variable-width bins of size 500kb and 1Mb
bins <- variableWidthBins(reads, binsizes=c(5e5,1e6))
## Plot the distribution of binsizes
hist(width(bins[['binsize_1e+06']]), breaks=50)
}
\author{
Aaron Taudt
}
