#' Bin the genome
#'
#' Please see functions \code{\link{fixedWidthBins}} and
#' \code{\link{variableWidthBins}} for further details.
#'
#' @name binning
NULL

#' Make fixed-width bins
#'
#' Make fixed-width bins based on given bin size.
#'
#' @param bamfile A BAM file from which the header is read to determine the
#'   chromosome lengths. If a \code{bamfile} is specified, option \code{assembly}
#'   is ignored.
#' @param assembly An assembly from which the chromosome lengths are
#'   determined. Please see
#'   \code{\link[GenomeInfoDb]{fetchExtendedChromInfoFromUCSC}} for available
#'   assemblies. This option is ignored if \code{bamfile} is specified.
#'   Alternatively a data.frame generated by
#'   \code{\link[GenomeInfoDb]{fetchExtendedChromInfoFromUCSC}}.
#' @param chrom.lengths A named character vector with chromosome lengths. Names correspond to chromosomes.
#' @param chromosome.format A character specifying the format of the
#'   chromosomes if \code{assembly} is specified. Either 'NCBI' for (1,2,3 ...)
#'   or 'UCSC' for (chr1,chr2,chr3 ...). If a \code{bamfile} or
#'   \code{chrom.lengths} is supplied, the format will be chosen automatically.
#' @param binsizes A vector of bin sizes in base pairs.
#' @param stepsizes A vector of step sizes in base pairs, the same length as \code{binsizes}.
#' @param chromosomes A subset of chromosomes for which the bins are generated.
#' @return A \code{list()} of \code{\link{GRanges-class}} objects with
#'   fixed-width bins. If \code{stepsizes} is specified, a \code{list()} of
#'   \code{\link{GRangesList}} objects with one entry per step.
#' @author Aaron Taudt
#' @importFrom Rsamtools BamFile
#' @export
#'
#' @examples
#' ## Make fixed-width bins of size 500kb and 1Mb
#' bins <- fixedWidthBins(assembly='mm10', chromosome.format='NCBI', binsizes=c(5e5,1e6))
#' bins
genomeBins <- function(seqinfo, binsize=1e6, reads.per.bin=NULL,
                            stepsize=NULL) {
    UseMethod("genomeBins")
}

genomeBins.Seqinfo <- function(seqinfo, binsize=1e6, reads.per.bin=NULL,
                                    stepsize=NULL) {
    if (is.null(stepsize))
        stepsize <- binsize
    if (! (binsize >= stepsize && binsize %% stepsize == 0))
        stop("binsize must be a multiple of stepsize")

    # fixed-size bins
    chr.len <- GenomeInfoDb::seqlengths(seqinfo)
    chr.len.floor <- floor(chr.len / stepsize) * stepsize
    # overlapping bins
    if (stepsize < binsize) {
        chr.len.floor <- chr.len.floor - (binsize/stepsize - 1) * stepsize
        chr.len.floor <- pmax(chr.len.floor, 0)
    }

    bins <- GenomicRanges::tileGenome(seqlengths=chr.len.floor, tilewidth=stepsize)
    bins <- unlist(bins, use.names=FALSE)
    if (stepsize < binsize)
        end(bins) <- (start(bins) + binsize - 1)

    GenomeInfoDb::seqinfo(bins) <- seqinfo
    attr(bins, 'binsize') <- binsize
    attr(bins, 'stepsize') <- stepsize
    bins
}

genomeBins.character <- function(seqinfo, ...) {
    genomeBins(genome(x), ...)
}

genomeBins.default <- function(x, ...) {
    stop("Do not know how to create bins from object type ", sQuote(class(x)))
}
